lib/elixir/lib/enum.ex

```diff
@@ -1941,39 +1977,46 @@ defmodule Enum do
 
   """
   @spec slice(t, integer, non_neg_integer) :: list
+  def slice(enumerable, start, n) when is_integer(start) and is_integer(n) and n >= 0,
+    do: do_slice(enumerable, start, n)
 
-  def slice(_enumerable, start, 0) when is_integer(start), do: []
+  defp do_slice(_enumerable, _start, 0),
+    do: []
 
-  def slice(enumerable, start, count)
-      when is_integer(start) and start < 0 and is_integer(count) and count >= 0 do
-    {list, new_start} = enumerate_and_count(enumerable, start)
-    if new_start >= 0 do
-      slice(list, new_start, count)
-    else
-      []
+  defp do_slice(enumerable, start, n) when start < 0 do
+    case get_size_index(enumerable, start) do
```

Comment:
There is an amount of duplication in every clause. I think this could be simpler if you change get_size_index to return `{enumerable, count}` no matter what. So the code would roughly look like this:

```elixir
{enumerable, count} = get_enumerable_count(enumerable)

with {:ok, start} when start < count <- normalize_index(index, count),
     {:ok, finish} <- normalize_index(start + n - 1, count) do
  do_slice(enumerable, {normalized_start, normalized_finish})
else
  :error -> []
end
```

where `get_enumerable_count(enumerable)` is:

```elixir
case Enumerable.count(enumerable) do
  {:ok, count} ->
    {enumerable, count}
  {:error, module} ->
    map_reduce(enumerable, 0, fn(x, acc) -> {x, acc + 1} end)
end
```

This way this function would be simpler and both `get_size_index` and `enumerate_count_index` would be gone.